src/Compilers/CSharp/Portable/Binder/Binder_Await.cs

```diff
@@ -307,51 +306,120 @@ private bool GetAwaitableExpressionInfo(
-                Debug.Assert(runtimeAwaitHelper.Arity == (resultType.HasType ? 1 : 0));
+                var awaitMembers = asyncHelpersType.GetMembers("Await");
+                runtimeAwaitHelper = null;
 
-                if (resultType.HasType)
+                foreach (var member in awaitMembers)
                 {
-                    runtimeAwaitHelper = runtimeAwaitHelper.Construct([resultType]);
-                    ConstraintsHelper.CheckConstraints(
-                        runtimeAwaitHelper,
-                        new ConstraintsHelper.CheckConstraintsArgs(this.Compilation, this.Conversions, includeNullability: false, expression.Syntax.Location, diagnostics));
+                    if (member is not MethodSymbol method
+                        || method.Arity != exprType.Arity
+                        || method.ParameterCount > 1)
+                    {
+                        continue;
+                    }
+
+                    if (method.Arity == 0)
+                    {
+                        if (method.ReturnsVoid && isValidConversion(exprType, method, node, diagnostics, this))
+                        {
+                            if (runtimeAwaitHelper is null)
+                            {
+                                runtimeAwaitHelper = method;
+                                continue;
+                            }
+                            else
+                            {
+                                runtimeAwaitHelper = null;
+                                return false;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        var unsubstitutedReturnType = method.ReturnType;
+                        if ((object)unsubstitutedReturnType != method.TypeArgumentsWithAnnotations[0].Type)
+                        {
+                            continue;
+                        }
+
+                        var substitutedMethod = method.Construct(exprType.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics);
+                        var tempDiagnostics = BindingDiagnosticBag.GetInstance(diagnostics);
+                        if (!ConstraintsHelper.CheckConstraints(
+                            substitutedMethod,
+                            new ConstraintsHelper.CheckConstraintsArgs(this.Compilation, this.Conversions, includeNullability: false, expression.Syntax.Location, tempDiagnostics)))
+                        {
+                            tempDiagnostics.Free();
+                            continue;
+                        }
+
+                        if (!isValidConversion(exprType, substitutedMethod, node, diagnostics, this))
+                        {
+                            tempDiagnostics.Free();
+                            continue;
+                        }
+
+                        diagnostics.AddRangeAndFree(tempDiagnostics);
+
+                        if (runtimeAwaitHelper is null)
```

Comment:
nit: the `tryGetRuntimeAwaitHelper` local function has a fair amount of complexity/depth. Consider breaking the contents of the loop out to another local function and factoring the ambiguity detection:
```
foreach (var member in awaitMembers)
{
  if (!isApplicableMember(member, out var candidate))
  {
    continue;
  }

  if (runtimeAwaitHelper is null)
  {
     runtimeAwaitHelper = candidate;
  }
  else
  {
    runtimeAwaitHelper = null;
    return false;
  }
}
```